[EMCAL-514] CCDB Container for the EMCAL bad channel map

mfasDa opened the pull request:
Container for the bad channels in EMCAL/DCAL stored in
the CCDB. Storing (index/status) of bad or warm cells.
Cell status can be checked with a dedicated getter,
where indices not stored in the container are
considered as good cells.

Additional functionality:
- Combination of bad channel map from different
  time stamps
- Comparison of different bad channel maps
  (must be identical in channel IDs and channel status)

For visualization a row-col representation of the
EMCAL bad channel map is created on-the-fly.

Implementation does not handle time stamps itself
but only the bad channel content - time stamps
are expected to be handled by general CCDB
functionality.

shahor02 commented on Detectors/EMCAL/calib/include/EMCALCalib/BadChannelMap.h:
```diff
@@ -0,0 +1,186 @@
+class BadChannelMap : public TObject
+{
+ public:
+  enum class MaskType_t {
```
Since you have just 3 values in the enum, why not to declare it as ``enum class MaskType_t : char {...`` to have the map more compact?

mfasDa replied:
Thanks a lot! Will fix this!

mfasDa commented:
Hi @shahor02, we have 3 possibilities for a channel (good, warm, bad), where a cell is automatically good in case it is not either warm or bad. We have in total 208*96 = 19968 cell indices (including empty indices in the PHOS region), so indeed two std::bitset<19968> would make the data more compact, faster and easier to handle. I will change this. Note that for the other containers (time / gain / temperature) we will have a floating point value (for time we try to limit short) per cell.  

Indeed I only inherit from TObject for the Browse function. We would like to have a possibility to visualise our container. The dynamic creation of a monitoring histogram works in any case, this is maybe sufficient and I can remove the inheritance from TObject.

In the interface I would like to stay with the enum class (changed as you suggested) to make the code more robust compared to using primitive data types encoding a channel status.

shahor02 commented:
> Note that for the other containers (time / gain / temperature) we will have a floating point value (for time we try to limit short) per cell.

OK, but I guess there every cell will have some value, no gain in using std::maps, plain std::array<float> will be more compact and faster.

mfasDa commented:
Of course, for the other objects every cell has a calibration value and we will use std::array.

mfasDa commented:
Before merging I would like to squash the commits first.

shahor02 commented on Detectors/EMCAL/calib/include/EMCALCalib/BadChannelMap.h:
```diff
@@ -0,0 +1,163 @@
+/// of the bad channel map from multiple time slices. For cells
+/// present in both bad channel maps always the worst case condition
+/// (BAD_CELL) is assumed.
+class BadChannelMap : public TObject
```
Still derives from TObject

mfasDa replied:
Inheritance removed.

shahor02 commented on Detectors/EMCAL/calib/include/EMCALCalib/BadChannelMap.h:
```diff
@@ -0,0 +1,163 @@
+  BadChannelMap() = default;
+
+  /// \brief Destructor
+  ~BadChannelMap() override = default;
```
once derivation from TObject is removed, remove also overrides

mfasDa replied:
Fixed

shahor02 commented on Detectors/EMCAL/calib/include/EMCALCalib/BadChannelMap.h:
```diff
@@ -0,0 +1,163 @@
+  std::bitset<19968>            mBadCells;          ///< Container for bad cells (size corresponding to the maximum amount of cells in EMCAL+DCAL including indices in the PHOS region)
+  std::bitset<19968>            mWarmCells;         ///< Contianer for warm cells (size corresponding to the maximum amount of cells in EMCAL+DCAL including indices in the PHOS region) 
+
+  ClassDefOverride(BadChannelMap, 1);
```
ClassDefNV(...

mfasDa replied:
Fixed

shahor02 commented on Detectors/EMCAL/calib/src/BadChannelMap.cxx:
```diff
@@ -0,0 +1,104 @@
+
+#include "FairLogger.h"
+
+#include <TBrowser.h>
```
not needed

mfasDa replied:
Fixed

shahor02 commented on Detectors/EMCAL/calib/src/BadChannelMap.cxx:
```diff
@@ -0,0 +1,104 @@
+
+#include <iostream>
+
+ClassImp(o2::emcal::BadChannelMap)
```
Also not needed

mfasDa replied:
Fixed

mfasDa commented:
Hi @shahor02 , 

Thanks for your comments! I removed the inheritance from TObject and fixed the formatting. The PR should be ready to merge. Please squash during merging!

Thanks a lot for your help!

Cheers

Markus

shahor02 commented:
@mfasDa Travis CI is still failing. Did you use clang-format 8 ?

mfasDa commented:
There was a change in the Geometry function which was not formatted - fixed. Hopefully that is the last formatting missing.

mfasDa commented:
One more change to distinguish dead cells (no data) from warm cells. The size of the container has also been adapted ignoring cell indices in the PHOS region - global col/row indices are now calculated based on the position in the supermodule in the histogram visualisation.

@shahor02 Anything still to be done before merging?

mfasDa commented:
Merging development PR with the PR for the corresponding unit test (coverage as described in commit message). The second pull request will be closed. All tests are successful - the code would be ready to merge.

shahor02 commented:
@mfasDa missed your last question, sorry. Everything looks fine, merging.
